File: kirun-java/src/test/java/com/fincity/nocode/kirun/engine/function/system/array/BinarySearchTest.java
package com.fincity.nocode.kirun.engine.function.system.array;

import java.util.Map;

import org.junit.jupiter.api.Test;

import com.fincity.nocode.kirun.engine.exception.KIRuntimeException;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveFunctionRepository;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveSchemaRepository;
import com.fincity.nocode.kirun.engine.runtime.reactive.ReactiveFunctionExecutionParameters;
import com.google.gson.JsonArray;
import com.google.gson.JsonNull;
import com.google.gson.JsonPrimitive;

import reactor.test.StepVerifier;

class BinarySearchTest {

	@Test
	void test() {
		var src = new JsonArray();
		src.add(2);
		src.add(4);
		src.add(10);
		src.add(12);
		src.add(20);
		src.add(1233);

		var search = new JsonPrimitive(20);

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
				.setArguments(Map.of(BinarySearch.PARAMETER_ARRAY_SOURCE_PRIMITIVE.getParameterName(), src,
						BinarySearch.PARAMETER_INT_SOURCE_FROM.getParameterName(), new JsonPrimitive(1),
						BinarySearch.PARAMETER_FIND_PRIMITIVE.getParameterName(), search,
						BinarySearch.PARAMETER_INT_LENGTH.getParameterName(), new JsonPrimitive(4)));
		BinarySearch bs = new BinarySearch();

		StepVerifier.create(bs.execute(fep))
				.expectNextMatches(result -> result.next().getResult().get(BinarySearch.EVENT_INDEX_NAME)
						.equals(new JsonPrimitive(4))).verifyComplete();
	}

	@Test
	void test2() {
		var src = new JsonArray();
		src.add(2);
		src.add(4);
		src.add(10);
		src.add(12);
		src.add(20);
		src.add(1233);

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
				.setArguments(Map.of(BinarySearch.PARAMETER_ARRAY_SOURCE_PRIMITIVE.getParameterName(), src,
						BinarySearch.PARAMETER_INT_SOURCE_FROM.getParameterName(), new JsonPrimitive(1),
						BinarySearch.PARAMETER_FIND_PRIMITIVE.getParameterName(), JsonNull.INSTANCE,
						BinarySearch.PARAMETER_INT_LENGTH.getParameterName(), new JsonPrimitive(4)));
		BinarySearch bs = new BinarySearch();

		StepVerifier.create(bs.execute(fep))
				.verifyError(KIRuntimeException.class);
	}

	@Test
	void test3() {

		JsonArray arr = new JsonArray();
		arr.add('a');
		arr.add('b');
		arr.add('c');
		arr.add('d');
		arr.add('e');
		arr.add('g');
		arr.add('i');
		arr.add('j');
		arr.add('k');
		arr.add('r');
		arr.add('s');
		arr.add('z');

		var res = new JsonPrimitive('c');

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
				.setArguments(Map.of(BinarySearch.PARAMETER_ARRAY_SOURCE_PRIMITIVE.getParameterName(), arr,
						BinarySearch.PARAMETER_INT_SOURCE_FROM.getParameterName(), new JsonPrimitive(1),
						BinarySearch.PARAMETER_FIND_PRIMITIVE.getParameterName(), res,
						BinarySearch.PARAMETER_INT_LENGTH.getParameterName(), new JsonPrimitive(8)));
		BinarySearch bs = new BinarySearch();

		StepVerifier.create(bs.execute(fep))
				.expectNextMatches(result -> result.next().getResult().get(BinarySearch.EVENT_INDEX_NAME)
						.equals(new JsonPrimitive(2))).verifyComplete();
	}

	@Test
	void test4() {

		JsonArray arr = new JsonArray();
		arr.add('a');
		arr.add('b');
		arr.add('c');
		arr.add('d');
		arr.add('e');
		arr.add('g');
		arr.add('i');
		arr.add('j');
		arr.add('k');
		arr.add('r');
		arr.add('s');
		arr.add('z');

		var res = new JsonPrimitive('z');

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
				.setArguments(Map.of(BinarySearch.PARAMETER_ARRAY_SOURCE_PRIMITIVE.getParameterName(), arr,
						BinarySearch.PARAMETER_INT_SOURCE_FROM.getParameterName(), new JsonPrimitive(-1),
						BinarySearch.PARAMETER_FIND_PRIMITIVE.getParameterName(), res,
						BinarySearch.PARAMETER_INT_LENGTH.getParameterName(), new JsonPrimitive(arr.size())));
		BinarySearch bs = new BinarySearch();

		StepVerifier.create(bs.execute(fep))
				.verifyError(KIRuntimeException.class);
	}

	@Test
	void test5() {

		JsonArray arr = new JsonArray();
		arr.add('a');
		arr.add('b');
		arr.add('c');
		arr.add('d');
		arr.add('e');
		arr.add('g');
		arr.add('i');
		arr.add('j');
		arr.add('k');
		arr.add('r');
		arr.add('s');
		arr.add('z');

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
				.setArguments(Map.of(BinarySearch.PARAMETER_ARRAY_SOURCE_PRIMITIVE.getParameterName(), arr,
						BinarySearch.PARAMETER_INT_SOURCE_FROM.getParameterName(), new JsonPrimitive(4),
						BinarySearch.PARAMETER_FIND_PRIMITIVE.getParameterName(), new JsonArray(),
						BinarySearch.PARAMETER_INT_LENGTH.getParameterName(), new JsonPrimitive(arr.size())));
		BinarySearch bs = new BinarySearch();

		StepVerifier.create(bs.execute(fep))
				.verifyError(KIRuntimeException.class);
	}
}

File: kirun-js/__tests__/engine/function/system/array/BinarySearchTest.ts
import { BinarySearch } from '../../../../../src/engine/function/system/array/BinarySearch';
import { FunctionExecutionParameters } from '../../../../../src/engine/runtime/FunctionExecutionParameters';
import { KIRunFunctionRepository, KIRunSchemaRepository } from '../../../../../src';

let bsearch: BinarySearch = new BinarySearch();

test('Binary Search test 1', async () => {
    let src: any[] = [1, 4, 6, 7, 10, 14, 16, 20];

    let search: any = 16;

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    ).setArguments(
        new Map<string, any>([
            [BinarySearch.PARAMETER_ARRAY_SOURCE_PRIMITIVE.getParameterName(), src],
            [BinarySearch.PARAMETER_INT_SOURCE_FROM.getParameterName(), 1],
            [BinarySearch.PARAMETER_FIND_PRIMITIVE.getParameterName(), search],
            [BinarySearch.PARAMETER_INT_LENGTH.getParameterName(), 6],
        ]),
    );

    expect(
        (await bsearch.execute(fep)).allResults()[0].getResult().get(BinarySearch.EVENT_INDEX_NAME),
    ).toBe(6);
});

test('Binary Search test 2', async () => {
    let src: any[] = [1, 4, 6, 7, 10, 14, 16, 20];

    let search: any = 78;

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    ).setArguments(
        new Map<string, any>([
            [BinarySearch.PARAMETER_ARRAY_SOURCE_PRIMITIVE.getParameterName(), src],
            [BinarySearch.PARAMETER_INT_SOURCE_FROM.getParameterName(), 1],
            [BinarySearch.PARAMETER_FIND_PRIMITIVE.getParameterName(), search],
            [BinarySearch.PARAMETER_INT_LENGTH.getParameterName(), src.length - 2],
        ]),
    );

    expect(
        (await bsearch.execute(fep)).allResults()[0].getResult().get(BinarySearch.EVENT_INDEX_NAME),
    ).toBe(-1);
});

test('Binary Search test 3', async () => {
    let src: any[] = [1, 4, 6, 7, 10, 14, 16, 20];

    let search: any = 78;

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    ).setArguments(
        new Map<string, any>([
            [BinarySearch.PARAMETER_ARRAY_SOURCE_PRIMITIVE.getParameterName(), src],
            [BinarySearch.PARAMETER_INT_SOURCE_FROM.getParameterName(), 1],
            [BinarySearch.PARAMETER_FIND_PRIMITIVE.getParameterName(), search],
            [BinarySearch.PARAMETER_INT_LENGTH.getParameterName(), 100],
        ]),
    );

    await expect(bsearch.execute(fep)).rejects.toThrow();
});

test('Binary Search test 6', async () => {
    let src: any[] = [1, 4, 6, 7, 10, 14, 17, 20];

    let search: number = 17;

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    ).setArguments(
        new Map<string, any>([
            [BinarySearch.PARAMETER_ARRAY_SOURCE_PRIMITIVE.getParameterName(), src],
            [BinarySearch.PARAMETER_INT_SOURCE_FROM.getParameterName(), 0],
            [BinarySearch.PARAMETER_FIND_PRIMITIVE.getParameterName(), search],
            [BinarySearch.PARAMETER_INT_LENGTH.getParameterName(), 5],
        ]),
    );

    expect(
        (await bsearch.execute(fep)).allResults()[0].getResult().get(BinarySearch.EVENT_INDEX_NAME),
    ).toBe(-1);
});

// test('Binary Search test 7', async () => {
//     let src: any[] = [1, 4, 6, 7, 10, 14, 16, 20];

//     let search: any[] = [10];

//     let fep: FunctionExecutionParameters = new FunctionExecutionParameters(new KIRunFunctionRepository(), new KIRunSchemaRepository()).setArguments(
//         new Map<string, any>([
//             [BinarySearch.PARAMETER_ARRAY_SOURCE_PRIMITIVE.getParameterName(), src],
//             [BinarySearch.PARAMETER_INT_SOURCE_FROM.getParameterName(), 1],
//             [BinarySearch.PARAMETER_FIND_PRIMITIVE.getParameterName(), search],
//             [BinarySearch.PARAMETER_INT_LENGTH.getParameterName(), 5],
//         ]),
//     );

//     expect(await bsearch.execute(fep)).toThrow();
// });

test('Binary Search test 4', async () => {
    let src: any[] = ['a', 'b', 'd', 'f', 'h', 'k', 'z'];

    let search: any = 'z';

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    ).setArguments(
        new Map<string, any>([
            [BinarySearch.PARAMETER_ARRAY_SOURCE_PRIMITIVE.getParameterName(), src],

            [BinarySearch.PARAMETER_FIND_PRIMITIVE.getParameterName(), search],
        ]),
    );

    expect(
        (await bsearch.execute(fep)).allResults()[0].getResult().get(BinarySearch.EVENT_INDEX_NAME),
    ).toBe(src.length - 1);
});

test('Binary Search test 5', async () => {
    let arr: any[] = [];
    arr.push('a');
    arr.push('b');
    arr.push('c');
    arr.push('d');
    arr.push('e');
    arr.push('g');
    arr.push('i');
    arr.push('j');
    arr.push('k');
    arr.push('r');
    arr.push('s');
    arr.push('z');

    let search: any = 's';

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    ).setArguments(
        new Map<string, any>([
            [BinarySearch.PARAMETER_ARRAY_SOURCE_PRIMITIVE.getParameterName(), arr],

            [BinarySearch.PARAMETER_FIND_PRIMITIVE.getParameterName(), search],
            [BinarySearch.PARAMETER_INT_LENGTH.getParameterName(), arr.length - 1],
        ]),
    );

    expect(
        (await bsearch.execute(fep)).allResults()[0].getResult().get(BinarySearch.EVENT_INDEX_NAME),
    ).toBe(10);
});
